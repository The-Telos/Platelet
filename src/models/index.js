// @ts-check
import { initSchema } from '@aws-amplify/datastore';
import { schema } from './schema';

const Role = {
  "USER": "USER",
  "COORDINATOR": "COORDINATOR",
  "RIDER": "RIDER",
  "ADMIN": "ADMIN"
};

const CommentVisibility = {
  "EVERYONE": "EVERYONE",
  "ME": "ME"
};

const Priority = {
  "HIGH": "HIGH",
  "MEDIUM": "MEDIUM",
  "LOW": "LOW"
};

const DeliverableTypeIcon = {
  "BUG": "BUG",
  "CHILD": "CHILD",
  "DOCUMENT": "DOCUMENT",
  "EQUIPMENT": "EQUIPMENT",
  "OTHER": "OTHER"
};

const DeliverableUnit = {
  "NONE": "NONE",
  "LITRE": "LITRE",
  "MILLILITRE": "MILLILITRE",
  "GRAM": "GRAM",
  "ITEM": "ITEM",
  "BOX": "BOX"
};

const TaskStatus = {
  "NEW": "NEW",
  "ACTIVE": "ACTIVE",
  "PICKED_UP": "PICKED_UP",
  "DROPPED_OFF": "DROPPED_OFF",
  "CANCELLED": "CANCELLED",
  "REJECTED": "REJECTED",
  "ABANDONED": "ABANDONED",
  "COMPLETED": "COMPLETED"
};

const Patch = {
  "NORTH": "NORTH",
  "WEST": "WEST",
  "EAST": "EAST",
  "SOUTH": "SOUTH",
  "RELIEF": "RELIEF",
  "AIR_AMBULANCE": "AIR_AMBULANCE"
};

const { User, AddressAndContactDetails, Vehicle, Comment, RiderResponsibility, Group, TaskAssignee, Task, Location, Deliverable, DeliverableType } = initSchema(schema);

export {
  User,
  AddressAndContactDetails,
  Vehicle,
  Comment,
  RiderResponsibility,
  Group,
  TaskAssignee,
  Task,
  Location,
  Deliverable,
  DeliverableType,
  Role,
  CommentVisibility,
  Priority,
  DeliverableTypeIcon,
  DeliverableUnit,
  TaskStatus,
  Patch
};