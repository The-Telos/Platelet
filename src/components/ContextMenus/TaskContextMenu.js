import React from "react";
import moment from "moment";
import Menu from "@mui/material/Menu";
import { makeStyles } from "@mui/material";
import MenuItem from "@mui/material/MenuItem";
import { useDispatch, useSelector } from "react-redux";
import MoreVertIcon from "@mui/icons-material/MoreVert";
import IconButton from "@mui/material/IconButton";
import { createPostingSelector } from "../../redux/LoadingSelectors";
import { deleteButtonStyles } from "./contextMenuCSS";
import PropTypes from "prop-types";
import * as models from "../../models/index";
import {
    displayErrorNotification,
    displayInfoNotification,
} from "../../redux/notifications/NotificationsActions";
import { DataStore } from "aws-amplify";

const initialState = {
    mouseX: null,
    mouseY: null,
};

function TaskContextMenu(props) {
    const dispatch = useDispatch();
    const { task } = props;
    const [state, setState] = React.useState(initialState);
    const deleteButtonClasses = deleteButtonStyles();
    const useStyles = makeStyles({
        button: {
            color: props.iconColor || "primary",
        },
    });
    const classes = useStyles();
    const postingSelector = createPostingSelector([
        "DELETE_TASK",
        "ADD_TASK_RELAY",
        "RESTORE_TASK",
        "UPDATE_TASK",
        "UPDATE_TASK_PICKUP_TIME",
        "UPDATE_TASK_DROPOFF_TIME",
        "UPDATE_TASK_CANCELLED_TIME",
        "UPDATE_TASK_REJECTED_TIME",
    ]);
    const isPosting = useSelector((state) => postingSelector(state));

    async function copyTaskDataToClipboard(e) {
        handleClose(e);
        if (!props.task || !props.task.id) {
            dispatch(displayErrorNotification("Copy failed."));
            return;
        }
        const taskResult = await DataStore.query(models.Task, props.task.id);
        if (taskResult) {
            const { pickUpLocation, priority, dropOffLocation, timeOfCall } =
                taskResult;
            const data = {
                FROM: pickUpLocation
                    ? `${pickUpLocation.ward || ""} - ${
                          pickUpLocation.line1 || ""
                      }`
                    : undefined,
                TO: dropOffLocation
                    ? `${dropOffLocation.ward || ""} - ${
                          dropOffLocation.line1 || ""
                      }`
                    : undefined,
                PRIORITY: priority || undefined,
                TOC: timeOfCall
                    ? moment(timeOfCall).format("HH:mm")
                    : undefined,
            };

            let result = "";
            let first = true;
            for (const [key, value] of Object.entries(data)) {
                if (value) result += `${first ? "" : " "}${key}: ${value}`;
                first = false;
            }

            navigator.clipboard.writeText(result).then(
                function () {
                    dispatch(displayInfoNotification("Copied to clipboard."));
                    /* clipboard successfully set */
                },
                function () {
                    dispatch(displayErrorNotification("Copy failed."));
                    /* clipboard write failed */
                }
            );
        } else {
            dispatch(displayErrorNotification("Copy failed."));
        }
    }

    const addRelay = (e) => {
        handleClose(e);
    };

    const handleClick = (event) => {
        event.preventDefault();
        setState({
            mouseX: event.clientX - 2,
            mouseY: event.clientY - 4,
        });
    };

    function onSelectPickedUp(e) {
        handleClose(e);
        const payload = new Date().toISOString();
        props.onSetTimePickedUp(payload);
        dispatch(
            displayInfoNotification("Task marked picked up", () => {
                props.onSetTimePickedUp(null);
            })
        );
    }

    function onSelectDroppedOff(e) {
        handleClose(e);
        const payload = new Date().toISOString();
        props.onSetTimeDroppedOff(payload);
        dispatch(
            displayInfoNotification("Task marked delivered", () => {
                props.onSetTimeDroppedOff(null);
            })
        );
    }
    function onSelectCancelled(e) {
        handleClose(e);
        const payload = new Date().toISOString();
        props.onSetTimeCancelled(payload);
        dispatch(
            displayInfoNotification("Task marked cancelled", () => {
                props.onSetTimeCancelled(null);
            })
        );
    }

    function onSelectRejected(e) {
        handleClose(e);
        const payload = new Date().toISOString();
        props.onSetTimeRejected(payload);
        dispatch(
            displayInfoNotification("Task marked rejected", () => {
                props.onSetTimeRejected(null);
            })
        );
    }

    function onDelete(e) {
        handleClose(e);
    }

    const handleClose = (e) => {
        e.preventDefault();
        setState(initialState);
    };

    return (
        <>
            <IconButton
                aria-label="more"
                aria-controls="long-menu"
                aria-haspopup="true"
                onClick={handleClick}
                disabled={isPosting}
            >
                <MoreVertIcon className={classes.button} />
            </IconButton>
            <Menu
                keepMounted
                open={state.mouseY !== null}
                onClose={handleClose}
                anchorReference="anchorPosition"
                anchorPosition={
                    state.mouseY !== null && state.mouseX !== null
                        ? { top: state.mouseY, left: state.mouseX }
                        : undefined
                }
            >
                <MenuItem
                    disabled={
                        task === null ||
                        !!task.timePickedUp ||
                        !props.assignedRiders.length > 0 ||
                        !!task.timeRejected ||
                        !!task.timeCancelled
                    }
                    onClick={onSelectPickedUp}
                >
                    Mark picked up
                </MenuItem>
                <MenuItem
                    disabled={
                        task === null ||
                        !!task.timeDroppedOff ||
                        !!!task.timePickedUp ||
                        !!task.timeRejected ||
                        !!task.timeCancelled
                    }
                    onClick={onSelectDroppedOff}
                >
                    Mark delivered
                </MenuItem>
                <MenuItem
                    disabled={
                        task === null ||
                        !!task.timeRejected ||
                        !!task.timeCancelled
                    }
                    onClick={onSelectRejected}
                >
                    Mark rejected
                </MenuItem>
                <MenuItem
                    disabled={
                        task === null ||
                        !!task.timeCancelled ||
                        !!task.timeRejected
                    }
                    onClick={onSelectCancelled}
                >
                    Mark cancelled
                </MenuItem>
                <MenuItem
                    disabled={
                        props.disableRelay ||
                        task === null ||
                        !!task.timeCancelled ||
                        !!task.timeRejected
                    }
                    onClick={addRelay}
                >
                    Add relay
                </MenuItem>
                <MenuItem onClick={copyTaskDataToClipboard}>
                    Save to clipboard
                </MenuItem>
                <MenuItem
                    className={
                        props.disableDeleted
                            ? deleteButtonClasses.deleteButtonDisabled
                            : deleteButtonClasses.deleteButton
                    }
                    onClick={onDelete}
                >
                    Delete
                </MenuItem>
            </Menu>
        </>
    );
}

TaskContextMenu.propTypes = {
    task: PropTypes.object,
    iconColor: PropTypes.string,
    disableDeleted: PropTypes.bool,
    disableRelay: PropTypes.bool,
    assignedRiders: PropTypes.arrayOf(PropTypes.object),
};

TaskContextMenu.defaultProps = {
    assignedRiders: [],
    task: null,
};

export default TaskContextMenu;
