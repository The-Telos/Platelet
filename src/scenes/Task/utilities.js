import { DataStore } from "aws-amplify";
import * as models from "../../models";
import determineTaskStatus from "../../utilities/determineTaskStatus";

export async function saveTaskTimeWithKey(key, value, taskId, taskAssignees) {
    let isoString = null;
    if (value) {
        isoString = new Date(value).toISOString();
    }
    const existingTask = await DataStore.query(models.Task, taskId);
    if (!existingTask) throw new Error("Task doesn't exist");
    const status = await determineTaskStatus(
        {
            ...existingTask,
            [key]: isoString,
        },
        taskAssignees.filter((ta) => ta.role === models.Role.RIDER)
    );
    return DataStore.save(
        models.Task.copyOf(existingTask, (updated) => {
            updated[key] = value ? isoString : null;
            updated.status = status;
        })
    );
}
