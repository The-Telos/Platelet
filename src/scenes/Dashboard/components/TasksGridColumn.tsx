import React, { useEffect, useRef, useState } from "react";
import _ from "lodash";
import * as models from "../../../models/index";
import { useSelector } from "react-redux";
import { Skeleton, Stack, useMediaQuery } from "@mui/material";
import { makeStyles } from "@mui/styles";
import {
    dashboardFilteredUserSelector,
    getRoleView,
    getWhoami,
    dashboardFilterTermSelector,
} from "../../../redux/Selectors";
import { filterTasks } from "../utilities/functions";
import GetError from "../../../ErrorComponents/GetError";
import Box from "@mui/material/Box";
import useWindowSize from "../../../hooks/useWindowSize";
import { useTheme } from "@mui/material/styles";
import TaskGridColumnHeader from "./TaskGridColumnHeader";
import TaskGridTasksList from "./TaskGridTasksList";
import useTasks from "../../../hooks/useTasks";

const useStyles = makeStyles((theme: any) => ({
    divider: {
        width: "95%",
    },
    spacer: {
        height: 35,
    },
    column: {
        padding: 5,
        backgroundColor: "rgba(180, 180, 180, 0.1)",
        borderRadius: 5,
        border: 0,
        boxShadow: "0 2px 3px 1px rgba(100, 100, 100, .3)",
        height: "100%",
        maxWidth: 360,
        minWidth: 285,
        [theme.breakpoints.down("lg")]: {
            padding: 0,
        },
        [theme.breakpoints.down("sm")]: {
            maxWidth: "100%",
            width: "100%",
        },
    },
    gridItem: {
        width: "100%",
    },
}));

type TasksGridColumnProps = {
    title?: string;
    taskKey: models.TaskStatus;
    showTasks?: string[];
};

const TasksGridColumn: React.FC<TasksGridColumnProps> = ({
    title = "TASKS",
    taskKey,
    showTasks = [],
}) => {
    const stateRef = useRef({});
    const [filteredTasksIds, setFilteredTasksIds] = useState<string[] | null>(
        null
    );
    const [width] = useWindowSize();
    const whoami = useSelector(getWhoami);
    const dashboardFilter = useSelector(dashboardFilterTermSelector);
    const dashboardFilteredUser = useSelector(dashboardFilteredUserSelector);
    const roleView = useSelector(getRoleView);

    const { state, isFetching, error } = useTasks(
        whoami.id,
        roleView,
        taskKey,
        dashboardFilteredUser
    );
    const tasks = Object.values(state);

    const theme = useTheme();

    const classes = useStyles();
    const isSm = useMediaQuery(theme.breakpoints.down("sm"));
    stateRef.current = state;

    /*function addTaskToState(newTask) {
        animate.current = true;
        setState((prevState) => ({
            ...prevState,
            [newTask.id]: newTask,
        }));
        animate.current = false;
    }

    function removeTaskFromState(newTask) {
        setState((prevState) => {
            if (prevState[newTask.id]) return _.omit(prevState, newTask.id);
            else return prevState;
        });
    }*/

    function doSearch() {
        const searchResult = filterTasks(state, dashboardFilter);
        setFilteredTasksIds(searchResult);
    }

    useEffect(doSearch, [dashboardFilter, state]);

    if (error) {
        return <GetError />;
    } else if (isFetching) {
        return (
            <Box
                data-testid="fetching-tasks-grid-column"
                className={classes.column}
                sx={{
                    width: isSm ? "100%" : width / 4.2,
                }}
            >
                <Stack direction="column" spacing={4}>
                    <Skeleton
                        variant="rectangular"
                        width={"100%"}
                        data-cy={`${title}-title-skeleton`}
                        height={50}
                    />
                    {_.range(4).map((i) => (
                        <Box key={i}>
                            <Skeleton
                                variant="rectangular"
                                width={"100%"}
                                height={200}
                            />
                        </Box>
                    ))}
                </Stack>
            </Box>
        );
    } else {
        return (
            <Box
                className={classes.column}
                sx={{
                    width: isSm ? "100%" : width / 4.2,
                }}
            >
                <TaskGridColumnHeader tasks={tasks} title={title} />
                <TaskGridTasksList
                    tasks={tasks}
                    includeList={filteredTasksIds}
                />
            </Box>
        );
    }
};

export default TasksGridColumn;
