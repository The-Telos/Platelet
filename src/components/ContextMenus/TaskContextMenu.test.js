import { render } from "../../test-utils";
import { mockAllIsIntersecting } from "react-intersection-observer/test-utils";
import TaskContextMenu from "./TaskContextMenu";
import { DataStore } from "aws-amplify";
import * as models from "../../models";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TasksGridColumn from "../../scenes/Dashboard/components/TasksGridColumn";
import * as copyTaskDataToClipboard from "../../utilities/copyTaskDataToClipboard";

const tenantId = "tenantId";

describe("TaskContextMenu", () => {
    const RealDate = Date;
    const isoDate = "2021-11-29T23:24:58.987Z";
    const dateString = "2021-11-29";

    function mockDate() {
        global.Date = class extends RealDate {
            constructor() {
                super();
                return new RealDate(isoDate);
            }
        };
    }

    afterEach(async () => {
        global.Date = RealDate;
    });
    beforeEach(async () => {
        jest.restoreAllMocks();
        mockDate();
    });

    test("cancel a task", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.NEW,
            })
        );
        const saveSpy = jest.spyOn(DataStore, "save");
        render(<TaskContextMenu task={task} assignedRiders={[]} />);
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Mark cancelled" })
        );
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeCancelled: isoDate,
                status: models.TaskStatus.CANCELLED,
            });
        });
        await waitFor(() => {
            expect(
                screen.getByText("Task marked cancelled")
            ).toBeInTheDocument();
        });
        // undo
        userEvent.click(screen.getByRole("button", { name: "UNDO" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeCancelled: null,
            });
        });
    });

    test("cancel a task failure", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.NEW,
            })
        );
        const saveSpy = jest
            .spyOn(DataStore, "save")
            .mockRejectedValue(new Error());
        render(<TaskContextMenu task={task} assignedRiders={[]} />);
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Mark cancelled" })
        );
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledTimes(1);
        });
        await waitFor(() => {
            expect(
                screen.getByText("Sorry, something went wrong")
            ).toBeInTheDocument();
        });
    });

    test("reject a task", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.NEW,
            })
        );
        const saveSpy = jest.spyOn(DataStore, "save");
        render(<TaskContextMenu task={task} assignedRiders={[]} />);
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Mark rejected" })
        );
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeRejected: isoDate,
                status: models.TaskStatus.REJECTED,
            });
        });
        await waitFor(() => {
            expect(
                screen.getByText("Task marked rejected")
            ).toBeInTheDocument();
        });
        // undo
        userEvent.click(screen.getByRole("button", { name: "UNDO" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeRejected: null,
            });
        });
    });

    test("mark a task picked up", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.ACTIVE,
            })
        );
        const fakeRider = await DataStore.save(
            new models.User({
                displayName: "Someone person",
                roles: [models.Role.USER, models.Role.RIDER],
            })
        );
        const fakeAssignment = await DataStore.save(
            new models.TaskAssignee({
                task,
                assignee: fakeRider,
                role: models.Role.RIDER,
            })
        );
        const preloadedState = {
            taskAssigneesReducer: {
                items: [fakeAssignment],
                ready: true,
                isSynced: true,
            },
        };
        const saveSpy = jest.spyOn(DataStore, "save");
        render(
            <TaskContextMenu task={task} assignedRiders={[fakeAssignment]} />,
            { preloadedState }
        );
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Mark picked up" })
        );
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timePickedUp: isoDate,
                status: models.TaskStatus.PICKED_UP,
            });
        });
        await waitFor(() => {
            expect(
                screen.getByText("Task marked picked up")
            ).toBeInTheDocument();
        });
        // undo
        userEvent.click(screen.getByRole("button", { name: "UNDO" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timePickedUp: null,
            });
        });
    });

    test("mark a task delivered", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.PICKED_UP,
                timePickedUp: new Date().toISOString(),
            })
        );
        const fakeRider = await DataStore.save(
            new models.User({
                displayName: "Someone person",
                roles: [models.Role.USER, models.Role.RIDER],
            })
        );
        const fakeAssignment = await DataStore.save(
            new models.TaskAssignee({
                task,
                assignee: fakeRider,
                role: models.Role.RIDER,
            })
        );
        const preloadedState = {
            taskAssigneesReducer: {
                items: [fakeAssignment],
                ready: true,
                isSynced: true,
            },
        };
        const saveSpy = jest.spyOn(DataStore, "save");
        render(
            <TaskContextMenu task={task} assignedRiders={[fakeAssignment]} />,
            { preloadedState }
        );
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Mark delivered" })
        );
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeDroppedOff: isoDate,
                status: models.TaskStatus.DROPPED_OFF,
            });
        });
        await waitFor(() => {
            expect(
                screen.getByText("Task marked delivered")
            ).toBeInTheDocument();
        });
        // undo
        userEvent.click(screen.getByRole("button", { name: "UNDO" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeDroppedOff: null,
            });
        });
    });
    test("mark a rider home", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.DROPPED_OFF,
                timePickedUp: new Date().toISOString(),
                timeDroppedOff: new Date().toISOString(),
            })
        );
        const fakeRider = await DataStore.save(
            new models.User({
                displayName: "Someone person",
                roles: [models.Role.USER, models.Role.RIDER],
            })
        );
        const fakeAssignment = await DataStore.save(
            new models.TaskAssignee({
                task,
                assignee: fakeRider,
                role: models.Role.RIDER,
            })
        );
        const preloadedState = {
            taskAssigneesReducer: {
                items: [fakeAssignment],
                ready: true,
                isSynced: true,
            },
        };
        const saveSpy = jest.spyOn(DataStore, "save");
        render(
            <TaskContextMenu task={task} assignedRiders={[fakeAssignment]} />,
            { preloadedState }
        );
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Mark rider home" })
        );
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeRiderHome: isoDate,
                status: models.TaskStatus.COMPLETED,
            });
        });
        await waitFor(() => {
            expect(
                screen.getByText("Task marked rider home")
            ).toBeInTheDocument();
        });
        // undo
        userEvent.click(screen.getByRole("button", { name: "UNDO" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...task,
                timeRiderHome: null,
            });
        });
    });
    test("copy to clipboard", async () => {
        Object.assign(window.navigator, {
            clipboard: {
                writeText: jest
                    .fn()
                    .mockImplementation(() => Promise.resolve()),
            },
        });

        const timeOfCall = "2022-09-14T07:55:07.473Z";
        const pickUpLocation = await DataStore.save(
            new models.Location({
                line1: "line one",
                ward: "test ward",
                postcode: "postcode",
            })
        );
        const dropOffLocation = await DataStore.save(
            new models.Location({
                line1: "something",
                ward: "some ward",
                postcode: "some postcode",
            })
        );
        const mockTask = new models.Task({
            timeOfCall,
            status: models.TaskStatus.NEW,
            pickUpLocation,
            dropOffLocation,
            priority: models.Priority.HIGH,
        });
        await DataStore.save(mockTask);
        const mockDeliverableType = new models.DeliverableType({
            label: "test deliverable",
        });
        await DataStore.save(mockDeliverableType);
        const mockDeliverables = [
            {
                unit: models.DeliverableUnit.ITEM,
                count: 1,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: models.DeliverableUnit.ITEM,
                count: 2,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: models.DeliverableUnit.LITER,
                count: 3,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
        ];
        await Promise.all(
            mockDeliverables.map((deliverable) =>
                DataStore.save(
                    new models.Deliverable({ task: mockTask, ...deliverable })
                )
            )
        );

        const copySpy = jest.spyOn(copyTaskDataToClipboard, "default");
        const clipboardSpy = jest.spyOn(navigator.clipboard, "writeText");
        render(<TaskContextMenu task={mockTask} assignedRiders={[]} />);
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(
            screen.getByRole("menuitem", { name: "Copy to clipboard" })
        );
        await waitFor(() => {
            expect(copySpy).toHaveBeenCalledWith(mockTask);
        });
        await waitFor(() => {
            expect(screen.getByText("Copied to clipboard")).toBeInTheDocument();
        });
        expect(clipboardSpy).toMatchSnapshot();
    });

    test.skip("duplicate a task", async () => {
        //TODO: check this
        const whoami = await DataStore.save(
            new models.User({
                displayName: "someone person",
                roles: [models.Role.USER, models.Role.COORDINATOR],
                tenantId,
            })
        );
        const mockLocation = await DataStore.save(
            new models.Location({
                name: "woop",
                listed: 1,
                tenantId,
            })
        );
        const mockLocation2 = await DataStore.save(
            new models.Location({
                name: "ohp",
                listed: 1,
                tenantId,
            })
        );
        const mockEstablishment = await DataStore.save(
            new models.Location({
                name: "something",
                listed: 1,
                tenantId,
            })
        );
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.NEW,
                pickUpLocation: mockLocation,
                dropOffLocation: mockLocation2,
                establishmentLocation: mockEstablishment,
                tenantId,
            })
        );
        const mockTaskAssignee = new models.TaskAssignee({
            task,
            assignee: whoami,
            role: models.Role.COORDINATOR,
            tenantId,
        });
        const deliverableTypes = await Promise.all(
            ["test deliverable", "another one"].map((d) =>
                DataStore.save(new models.DeliverableType({ label: d }))
            )
        );
        const deliverables = await Promise.all(
            deliverableTypes.map((deliverableType) =>
                DataStore.save(
                    new models.Deliverable({
                        deliverableType,
                        count: 3,
                        unit: deliverableType.defaultUnit,
                        task,
                        tenantId,
                    })
                )
            )
        );
        const saveSpy = jest.spyOn(DataStore, "save");
        const preloadedState = {
            whoami: { user: whoami },
            roleView: models.Role.COORDINATOR,
            taskAssigneesReducer: {
                items: [mockTaskAssignee],
                ready: true,
                isSynced: true,
            },
            tenantId,
        };
        //render(<TaskContextMenu task={task} assignedRiders={[]} />, {
        //    preloadedState,
        //});
        render(
            <>
                <TaskContextMenu task={task} assignedRiders={[]} />
                <TasksGridColumn taskKey={models.TaskStatus.NEW} />
            </>,
            {
                preloadedState,
            }
        );

        await waitFor(() => {
            expect(
                screen.queryByTestId("fetching-tasks-grid-column")
            ).toBeNull();
        });
        mockAllIsIntersecting(true);
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(screen.getByRole("menuitem", { name: "Duplicate" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledTimes(4);
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...task,
            dateCreated: dateString,
            userCreatedTasksId: whoami.id,
            tenantId,
            id: expect.not.stringMatching(task.id),
        });
        deliverables.forEach((del) => {
            expect(saveSpy).toHaveBeenCalledWith({
                ...del,
                tenantId,
                id: expect.not.stringMatching(del.id),
                taskDeliverablesId: expect.not.stringMatching(task.id),
            });
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...mockTaskAssignee,
            tenantId,
            id: expect.any(String),
            taskAssigneesId: expect.not.stringMatching(task.id),
        });
        await screen.findByText("Task duplicated to NEW");
        mockAllIsIntersecting(true);
        await waitFor(() => {
            expect(screen.queryByTestId("task-item-skeleton")).toBeNull();
        });
        await waitFor(
            () => {
                expect(screen.queryAllByRole("link")).toHaveLength(2);
            },
            { timeout: 10000 }
        );
    });

    test("can't duplicate in rider role view", async () => {
        const whoami = await DataStore.save(
            new models.User({
                displayName: "someone person",
                roles: [models.Role.USER, models.Role.RIDER],
            })
        );
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.ACTIVE,
                riderResponsibility: "something",
            })
        );
        const preloadedState = {
            whoami: { user: whoami },
            roleView: models.Role.RIDER,
        };
        render(<TaskContextMenu task={task} assignedRiders={[]} />, {
            preloadedState,
        });
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        expect(
            screen.queryByRole("menuitem", { name: "Duplicate" })
        ).toBeNull();
    });

    test.skip("duplicate a task in rider role view", async () => {
        // at the moment riders can't duplicate tasks
        const whoami = await DataStore.save(
            new models.User({
                displayName: "someone person",
                roles: [models.Role.USER, models.Role.RIDER],
                tenantId,
            })
        );
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.ACTIVE,
                riderResponsibility: "something",
            })
        );
        const mockTaskAssignee = new models.TaskAssignee({
            task,
            assignee: whoami,
            role: models.Role.RIDER,
            tenantId: whoami.tenantId,
        });
        const saveSpy = jest.spyOn(DataStore, "save");
        const preloadedState = {
            whoami: { user: whoami },
            roleView: models.Role.RIDER,
        };
        render(<TaskContextMenu task={task} assignedRiders={[]} />, {
            preloadedState,
        });
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(screen.getByRole("menuitem", { name: "Duplicate" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledTimes(2);
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...task,
            id: expect.not.stringMatching(task.id),
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...mockTaskAssignee,
            id: expect.any(String),
            task: {
                ...task,
                id: expect.any(String),
            },
        });
    });

    test("duplicate a task with unlisted locations", async () => {
        const whoami = await DataStore.save(
            new models.User({
                displayName: "someone person",
                roles: [models.Role.USER, models.Role.COORDINATOR],
                tenantId,
            })
        );
        const mockLocation = await DataStore.save(
            new models.Location({ name: "woop", listed: 0, tenantId })
        );
        const mockLocation2 = await DataStore.save(
            new models.Location({ name: "ohp", listed: 0, tenantId })
        );
        const mockEstablishment = await DataStore.save(
            new models.Location({ name: "something", listed: 0, tenantId })
        );
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.NEW,
                pickUpLocation: mockLocation,
                dropOffLocation: mockLocation2,
                establishmentLocation: mockEstablishment,
                tenantId,
            })
        );
        const saveSpy = jest.spyOn(DataStore, "save");
        const preloadedState = {
            whoami: { user: whoami },
            tenantId,
        };
        render(<TaskContextMenu task={task} assignedRiders={[]} />, {
            preloadedState,
        });
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(screen.getByRole("menuitem", { name: "Duplicate" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledTimes(5);
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...mockLocation,
            id: expect.not.stringMatching(mockLocation.id),
            tenantId,
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...mockLocation2,
            id: expect.not.stringMatching(mockLocation2.id),
            tenantId,
        });
        expect(saveSpy).toHaveBeenCalledWith({
            ...mockEstablishment,
            id: expect.not.stringMatching(mockEstablishment.id),
            tenantId,
        });
    });

    test("duplicate a task failure", async () => {
        const task = await DataStore.save(
            new models.Task({
                status: models.TaskStatus.NEW,
            })
        );
        const whoami = await DataStore.save(
            new models.User({
                displayName: "someone person",
                roles: [models.Role.USER, models.Role.COORDINATOR],
                tenantId,
            })
        );
        const saveSpy = jest
            .spyOn(DataStore, "save")
            .mockRejectedValue(new Error());
        const preloadedState = {
            whoami: { user: whoami },
            tenantId,
        };
        render(<TaskContextMenu task={task} assignedRiders={[]} />, {
            preloadedState,
        });
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        userEvent.click(screen.getByRole("menuitem", { name: "Duplicate" }));
        await waitFor(() => {
            expect(saveSpy).toHaveBeenCalledTimes(1);
        });
        await waitFor(() => {
            expect(
                screen.getByText("Sorry, something went wrong")
            ).toBeInTheDocument();
        });
    });

    test.each`
        taskStatus
        ${models.TaskStatus.COMPLETED} | ${models.TaskStatus.ABANDONED} | ${models.TaskStatus.REJECTED} | ${models.TaskStatus.CANCELLED}
    `("disable duplicate if completed", async ({ taskStatus }) => {
        const task = await DataStore.save(
            new models.Task({
                status: taskStatus,
            })
        );
        render(<TaskContextMenu task={task} assignedRiders={[]} />);
        const button = screen.getByRole("button", { name: "task options" });
        userEvent.click(button);
        expect(
            screen.getByRole("menuitem", { name: "Duplicate" })
        ).toHaveAttribute("aria-disabled", "true");
    });
});
