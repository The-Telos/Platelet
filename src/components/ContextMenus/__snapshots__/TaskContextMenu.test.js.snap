// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskContextMenu copy to clipboard 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "TOC: 23:24 FROM: test ward - line one, postcode TO: some ward - something, some postcode PRIORITY: high ITEMS: test deliverable x 1, test deliverable x 2, test deliverable x 3",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
